In [30]:
import pandas as pd
data = pd.read_csv('prostate.txt')
data 

lcavol   lweight  age      lbph  svi       lcp  gleason  pgg45  \
0  -0.579818  2.769459   50 -1.386294    0 -1.386294        6      0   
1  -0.994252  3.319626   58 -1.386294    0 -1.386294        6      0   
2  -0.510826  2.691243   74 -1.386294    0 -1.386294        7     20   
3  -1.203973  3.282789   58 -1.386294    0 -1.386294        6      0   
4   0.751416  3.432373   62 -1.386294    0 -1.386294        6      0   
..       ...       ...  ...       ...  ...       ...      ...    ...   
92  2.830268  3.876396   68 -1.386294    1  1.321756        7     60   
93  3.821004  3.896909   44 -1.386294    1  2.169054        7     40   
94  2.907447  3.396185   52 -1.386294    1  2.463853        7     10   
95  2.882564  3.773910   68  1.558145    1  1.558145        7     80   
96  3.471966  3.974998   68  0.438255    1  2.904165        7     20   

        lpsa  train  
0  -0.430783   True  
1  -0.162519   True  
2  -0.162519   True  
3  -0.162519   True  
4   0.371564   True  
..       ...    ...  
92  4.385147   True  
93  4.684443   True  
94  5.143124  False  
95  5.477509   True  
96  5.582932  False  

[97 rows x 10 columns]

In [31]:
data.info()
data.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 97 entries, 0 to 96
Data columns (total 10 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   lcavol   97 non-null     float64
 1   lweight  97 non-null     float64
 2   age      97 non-null     int64  
 3   lbph     97 non-null     float64
 4   svi      97 non-null     int64  
 5   lcp      97 non-null     float64
 6   gleason  97 non-null     int64  
 7   pgg45    97 non-null     int64  
 8   lpsa     97 non-null     float64
 9   train    97 non-null     bool   
dtypes: bool(1), float64(5), int64(4)
memory usage: 7.0 KB


lcavol    lweight        age       lbph        svi        lcp  \
count  97.000000  97.000000  97.000000  97.000000  97.000000  97.000000   
mean    1.350010   3.628943  63.865979   0.100356   0.216495  -0.179366   
std     1.178625   0.428411   7.445117   1.450807   0.413995   1.398250   
min    -1.347074   2.374906  41.000000  -1.386294   0.000000  -1.386294   
25%     0.512824   3.375880  60.000000  -1.386294   0.000000  -1.386294   
50%     1.446919   3.623007  65.000000   0.300105   0.000000  -0.798508   
75%     2.127041   3.876396  68.000000   1.558145   0.000000   1.178655   
max     3.821004   4.780383  79.000000   2.326302   1.000000   2.904165   

         gleason       pgg45       lpsa  
count  97.000000   97.000000  97.000000  
mean    6.752577   24.381443   2.478387  
std     0.722134   28.204035   1.154329  
min     6.000000    0.000000  -0.430783  
25%     6.000000    0.000000   1.731656  
50%     7.000000   15.000000   2.591516  
75%     7.000000   40.000000   3.056357  
max     9.000000  100.000000   5.582932

In [32]:
import pandas as pd
data = pd.read_csv('prostate.txt')
data

lcavol   lweight  age      lbph  svi       lcp  gleason  pgg45  \
0  -0.579818  2.769459   50 -1.386294    0 -1.386294        6      0   
1  -0.994252  3.319626   58 -1.386294    0 -1.386294        6      0   
2  -0.510826  2.691243   74 -1.386294    0 -1.386294        7     20   
3  -1.203973  3.282789   58 -1.386294    0 -1.386294        6      0   
4   0.751416  3.432373   62 -1.386294    0 -1.386294        6      0   
..       ...       ...  ...       ...  ...       ...      ...    ...   
92  2.830268  3.876396   68 -1.386294    1  1.321756        7     60   
93  3.821004  3.896909   44 -1.386294    1  2.169054        7     40   
94  2.907447  3.396185   52 -1.386294    1  2.463853        7     10   
95  2.882564  3.773910   68  1.558145    1  1.558145        7     80   
96  3.471966  3.974998   68  0.438255    1  2.904165        7     20   

        lpsa  train  
0  -0.430783   True  
1  -0.162519   True  
2  -0.162519   True  
3  -0.162519   True  
4   0.371564   True  
..       ...    ...  
92  4.385147   True  
93  4.684443   True  
94  5.143124  False  
95  5.477509   True  
96  5.582932  False  

[97 rows x 10 columns]

In [33]:
# Avval mavjud ustunlarni aniqlaymiz
cols_to_drop = [col for col in ['id', 'lpsa', 'train'] if col in data.columns]

# Ustunlarni tushiramiz
X = data.drop(cols_to_drop, axis='columns')

# Agar 'lpsa' mavjud bo‘lsa, uni y deb ajratamiz
if 'lpsa' in data.columns:
    y = data['lpsa']
else:
    raise ValueError("'lpsa' ustuni DataFrame ichida topilmadi!")

# Train/Test ga ajratamiz
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=55)

# Natijani chiqaramiz
print(X_train.shape, X_test.shape)



(77, 8) (20, 8)


In [34]:
print(data.columns)


Index(['lcavol', 'lweight', 'age', 'lbph', 'svi', 'lcp', 'gleason', 'pgg45',
       'lpsa', 'train'],
      dtype='object')


In [35]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 97 entries, 0 to 96
Data columns (total 10 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   lcavol   97 non-null     float64
 1   lweight  97 non-null     float64
 2   age      97 non-null     int64  
 3   lbph     97 non-null     float64
 4   svi      97 non-null     int64  
 5   lcp      97 non-null     float64
 6   gleason  97 non-null     int64  
 7   pgg45    97 non-null     int64  
 8   lpsa     97 non-null     float64
 9   train    97 non-null     bool   
dtypes: bool(1), float64(5), int64(4)
memory usage: 7.0 KB


0 - > row
1 - > collumns
 lavel ground truth

In [36]:

cols_to_drop = [col for col in ['id', 'lpsa', 'train'] if col in data.columns]
X = data.drop(cols_to_drop, axis='columns')

if 'lpsa' in data.columns:
    y = data['lpsa']
else:
    raise KeyError("'lpsa' ustuni topilmadi!")
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=55)
print(X_train.shape, X_test.shape)



(77, 8) (20, 8)


In [37]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)
print(X_train.shape, X_test.shape)

model = LinearRegression()
model.fit(X_train, y_train)

y_pred = model.predict(X_test)
mse = mean_squared_error(y_test, y_pred)
mae = mean_absolute_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f"Mean Squared Error: {mse}")
print(f"Mean Absolute Error: {mae}")
print(f"R2 Score: {r2}")

(67, 8) (30, 8)
Mean Squared Error: 0.4612959596185864
Mean Absolute Error: 0.530383981112604
R2 Score: 0.6039185108581981


In [38]:
from sklearn.linear_model import Ridge

ridge_model = Ridge(alpha=10)
ridge_model.fit(X_train, y_train)


y_pred = ridge_model.predict(X_test)
mse = mean_squared_error(y_test, y_pred)
mae = mean_absolute_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f"Mean Squared Error: {mse}")
print(f"Mean Absolute Error: {mae}")
print(f"R2 Score: {r2}")

Mean Squared Error: 0.4451186228892906
Mean Absolute Error: 0.49290134977139155
R2 Score: 0.6178088202972527


In [39]:
from sklearn.linear_model import Lasso

lasso_model = Lasso(alpha=0.1)
lasso_model.fit(X_train, y_train)


y_pred = lasso_model.predict(X_test)
mse = mean_squared_error(y_test, y_pred)
mae = mean_absolute_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f"Mean Squared Error: {mse}")
print(f"Mean Absolute Error: {mae}")
print(f"R2 Score: {r2}")

Mean Squared Error: 0.5404574731660402
Mean Absolute Error: 0.5451558403532795
R2 Score: 0.5359482424983382


### HYPERPARAMTER tuning
Ridge, alpha ,1 ,10, 100


Natijalar:

BAse model (Simple Linnear Regressions): 0.6039185108581981
Ridge(alpha=15) : 0.6178088202972527
Lasso(alpha=0.01) :0.5359482424983382

### Data processing - normalitions

x_scaled = (x-x_min)/(x_max-x-min)

In [40]:
import pandas as pd
data = pd.read_csv('prostate.txt')
data 
cols_to_drop = [col for col in ['id', 'lpsa', 'train'] if col in data.columns]
X = data.drop(cols_to_drop, axis='columns')
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)


from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()

scaler.fit(X_train)
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)


In [41]:
from sklearn.linear_model import Ridge,Lasso, LinearRegression
from sklearn.metrics import mean_absolute_error,mean_squared_error, r2_score

In [42]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score

# Ma'lumotni o'qish
data = pd.read_csv('prostate.txt')

# Target ustunni ajratish
y = data['lpsa']

# ID va train ustunlarini olib tashlash
cols_to_drop = [col for col in ['id', 'lpsa', 'train'] if col in data.columns]
X = data.drop(cols_to_drop, axis='columns')

# Train-test bo'lish
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Skalirlash
scaler = MinMaxScaler()
scaler.fit(X_train)
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Model qurish
model = LinearRegression()
model.fit(X_train_scaled, y_train)

# Bashorat va metrikalar
y_pred = model.predict(X_test_scaled)
mse = mean_squared_error(y_test, y_pred)
mae = mean_absolute_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f"\nMean Squared Error: {mse}")
print(f"Mean Absolute Error: {mae}")
print(f"R2 Score: {r2}")



Mean Squared Error: 0.46129595961858627
Mean Absolute Error: 0.5303839811126044
R2 Score: 0.6039185108581981


In [43]:
from sklearn.linear_model import Ridge
ridge_model = Ridge()
ridge_model.fit(X_train_scaled, y_train)

# Bashorat va metrikalar
y_pred = ridge_model.predict(X_test_scaled)
mse = mean_squared_error(y_test, y_pred)
mae = mean_absolute_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f"\nMean Squared Error: {mse}")
print(f"Mean Absolute Error: {mae}")
print(f"R2 Score: {r2}")


Mean Squared Error: 0.4158121965885847
Mean Absolute Error: 0.5015838360115886
R2 Score: 0.6429721297270723


In [44]:
from sklearn.linear_model import Lasso
lasso_model = Lasso(alpha=0.01)
lasso_model.fit(X_train_scaled, y_train)

# Bashorat va metrikalar
y_pred = lasso_model.predict(X_test_scaled)
mse = mean_squared_error(y_test, y_pred)
mae = mean_absolute_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f"\nMean Squared Error: {mse}")
print(f"Mean Absolute Error: {mae}")
print(f"R2 Score: {r2}")


Mean Squared Error: 0.4416049725137688
Mean Absolute Error: 0.5216348696806314
R2 Score: 0.620825737840192


In [45]:
X_train.head(1)

lcavol  lweight  age      lbph  svi       lcp  gleason  pgg45
95  2.882564  3.77391   68  1.558145    1  1.558145        7     80

In [46]:
X_train_scaled[0]

array([0.87394984, 0.58159109, 0.69444444, 0.79309437, 1.        ,
       0.68627592, 0.33333333, 0.8       ])

In [47]:
a = (7-6)/(9-6)
a

0.3333333333333333

### Data preprocessing -standardization

x_scaled = (x - mean)/std

In [51]:
import pandas as pd
from sklearn.linear_model import Ridge, Lasso, LinearRegression
from sklearn.preprocessing import StandardScaler, MinMaxScaler
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score


data = pd.read_csv('prostate.txt')
X = data.drop(['lpsa', 'train'], axis='columns')
y = data['lpsa']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
print(X_train_scaled, X_test.shape)


scaler = StandardScaler()
scaler.fit(X_train)
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)


# lasso_model = Lasso(alpha=0.01)
# lasso_model.fit(X_train_scaled, y_train)

ridge_model = Ridge(alpha=10)
ridge_model.fit(X_train_scaled, y_train)

# Bashorat va metrikalar
y_pred = ridge_model.predict(X_test_scaled)
mse = mean_squared_error(y_test, y_pred)
mae = mean_absolute_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f"\nMean Squared Error: {mse}")
print(f"Mean Absolute Error: {mae}")
print(f"R2 Score: {r2}")



[[-0.12469333 -0.21218779  0.78741288  0.80598507 -0.53229065 -0.42892998
  -1.21976137 -0.84401113]
 [ 0.43355199 -0.15577023 -0.63251199 -1.07345361  1.87867287  1.10500403
   0.55234477  1.52640311]
 [-1.57150246  1.67902815  0.64542039 -1.07345361 -0.53229065 -0.85417238
  -1.21976137 -0.84401113]
 [ 0.15330463 -1.35325558  0.21944293 -1.07345361 -0.53229065 -0.85417238
  -1.21976137 -0.84401113]
 [-0.49207437 -1.46120529 -1.20048193  0.50798465 -0.53229065  0.0311849
   0.55234477 -0.64647661]
 [-1.17555636 -0.88830969 -1.4844669  -1.07345361 -0.53229065 -0.85417238
  -1.21976137 -0.84401113]
 [-0.99053642 -0.94034139 -0.20653453 -1.07345361 -0.53229065 -0.85417238
  -1.21976137 -0.84401113]
 [ 0.08970442 -1.33227368  0.21944293  0.25162089 -0.53229065  0.31019776
   0.55234477  0.73626503]
 [-2.35325094 -0.14342277 -0.20653453  0.73328103 -0.53229065 -0.85417238
  -1.21976137 -0.84401113]
 [ 0.0816687   0.39111627 -0.4905195   0.7706029  -0.53229065 -0.1628944
   0.55234477 -0.05

1. Linnear Regression model type (turi)
 - linnear regression 
 - Lasso
 - Ridge

2. Hyperparamater tunning
 - Ridge alpha (10,1) vs hz .. 
 - Lasso alpha (0.01) . ..

3. Data preprocessing
 - Minmaxscaler c scale = (x-x_min)/(x_max-x_min)
 - Standart scaler x_scaled = (x-mean)/std

4. Data splitting:
 - Nechchi % ini testga berish
 - Qanday random value ishlatish
 - 0.2215489754484546848

In [49]:
data

lcavol   lweight  age      lbph  svi       lcp  gleason  pgg45  \
0  -0.579818  2.769459   50 -1.386294    0 -1.386294        6      0   
1  -0.994252  3.319626   58 -1.386294    0 -1.386294        6      0   
2  -0.510826  2.691243   74 -1.386294    0 -1.386294        7     20   
3  -1.203973  3.282789   58 -1.386294    0 -1.386294        6      0   
4   0.751416  3.432373   62 -1.386294    0 -1.386294        6      0   
..       ...       ...  ...       ...  ...       ...      ...    ...   
92  2.830268  3.876396   68 -1.386294    1  1.321756        7     60   
93  3.821004  3.896909   44 -1.386294    1  2.169054        7     40   
94  2.907447  3.396185   52 -1.386294    1  2.463853        7     10   
95  2.882564  3.773910   68  1.558145    1  1.558145        7     80   
96  3.471966  3.974998   68  0.438255    1  2.904165        7     20   

        lpsa  train  
0  -0.430783   True  
1  -0.162519   True  
2  -0.162519   True  
3  -0.162519   True  
4   0.371564   True  
..       ...    ...  
92  4.385147   True  
93  4.684443   True  
94  5.143124  False  
95  5.477509   True  
96  5.582932  False  

[97 rows x 10 columns]